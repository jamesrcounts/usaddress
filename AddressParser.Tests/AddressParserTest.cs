using System.IO;
using System;
using System.Linq;
using System.Text.RegularExpressions;

using ApprovalTests;
using ApprovalTests.Combinations;

using ApprovalUtilities.Utilities;

using Xunit;

namespace USAddress.Tests
{
    public class AddressParserTest
    {
        private static readonly AddressParser Parser = AddressParser.Default;

        [Fact]
        public void StreetLineTests()
        {
            var addressParser = new AddressParser();
            var streetPattern = new Regex(addressParser.StreetPattern, addressParser.MatchOptions);
            var streetPatternMatch = streetPattern.Match("COUNTY ROAD F");
            Assert.NotNull(streetPatternMatch);
            Assert.True(streetPatternMatch.Success);
            Assert.True(streetPatternMatch.Captures.Count == 1);
            Assert.True(streetPatternMatch.Captures[0].Value == "COUNTY ROAD F");
        }

        [Fact]
        public void ParseAddressLine()
        {
            var addressParser = new AddressParser();
            var result = addressParser.ParseAddressLine("3360 County Road F", true);
            Approvals.Verify(result);
        }

        [Fact]
        public void StateDoesNotMatchCity()
        {
            var reg = new Regex(Parser.PlacePattern, Parser.MatchOptions);
            var match = reg.Match("URB LA FABRICA,PR,00704");
            var state = match.Groups[Components.State].Value;
            Approvals.Verify(state);
        }

        [Fact]
        public void AddressRegexIsLazy()
        {
            var addressParser = new AddressParser();
            addressParser.StreetSuffixes.Add("EDGE", "EDGE");
            Approvals.Verify(addressParser.AddressRegex);
        }

        /// <summary>
        /// People in Wisconsin and Illinois are eating too much cheese, apparently, because
        /// you can encounter house numbers with letters in them. It's similar to the
        /// Utah grid-system, except the gridness is all crammed into the house number.
        /// </summary>
        [Fact]
        public void CanParseAddressWithAlphanumericRange()
        {
            Approvals.Verify(Parser.ParseAddress("N6W23001 BLUEMOUND ROAD, ROLLING MEADOWS, IL, 12345"));
        }

        /// <summary>
        /// In Virginia Beach, for example, there's a South Blvd, which could really
        /// throw a spanner into our predirectional/postdirectional parsing. We call
        /// this case out specifically in our regex.
        /// </summary>
        [Fact]
        public void CanParseAddressWithCardinalStreetName()
        {
            Approvals.Verify(Parser.ParseAddress("500 SOUTH STREET VIRGINIA BEACH VIRGINIA 23452"));
        }

        [Fact]
        public void CanParseAddressWithoutPunctuation()
        {
            Approvals.Verify(Parser.ParseAddress("999 West 89th Street Apt A New York NY 10024"));
        }

        /// <summary>
        /// When people live in apartments with letters, they sometimes attach the apartment
        /// letter to the end of the house number. This is wrong, and these people need to be
        /// lined up and individually slapped. We pull out the unit and designate it as "APT",
        /// which in my experience is the designator that USPS uses in the vast, vast majority
        /// of cases.
        /// </summary>
        [Fact]
        public void CanParseAddressWithRangedUnitAttachedToNumber()
        {
            Approvals.Verify(Parser.ParseAddress("403D BERRYFIELD LANE CHESAPEAKE VA 23224"));
        }

        [Fact]
        public void CanParseAddressWithRangelessSecondaryUnit()
        {
            Approvals.Verify(Parser.ParseAddress("1050 Broadway Penthouse, New York, NY 10001"));
        }

        /// <summary>
        /// Speaking of weird addresses, sometimes people put a space in the number.
        /// USPS says we should squash it together.
        /// </summary>
        [Fact]
        public void CanParseAddressWithSpacedAlphanumericRange()
        {
            Approvals.Verify(Parser.ParseAddress("N645 W23001 BLUEMOUND ROAD, ROLLING MEADOWS, IL, 12345"));
        }

        /// <summary>
        /// At least it's not platform 9 3/4.
        /// </summary>
        [Fact]
        public void CanParseFractionalAddress()
        {
            Approvals.Verify(Parser.ParseAddress("123 1/2 MAIN ST, RICHMOND, VA 23221"));
        }

        /// <summary>
        /// Grid-style addresses are common in parts of Utah. The official USPS address database
        /// in this case treats "E" as a pre-directional, "1700" as the street name, and "S" as a
        /// post-directional, and nothing as the suffix, so that's how we parse it, too.
        /// </summary>
        [Fact]
        public void CanParseGridStyleAddress()
        {
            Approvals.Verify(Parser.ParseAddress("842 E 1700 S, Salt Lake City, UT 84105"));
        }

        /// <summary>
        /// Military addresses seem to follow no convention whatsoever in the
        /// street line, but the APO/FPO/DPO AA/AE/AP 9NNNN part of the place line
        /// is pretty well standardized. I've made a special exception for these
        /// kinds of addresses so that the street line is just dumped as-is into
        /// the StreetLine field.
        /// </summary>
        [Fact]
        public void CanParseMilitaryAddress()
        {
            Approvals.Verify(Parser.ParseAddress("PSC BOX 453, APO AE 99969"));
        }

        [Fact]
        public void CanParsePostOfficeBoxAddress()
        {
            Approvals.Verify(Parser.ParseAddress("P.O. BOX 4857, New York, NY 10001"));
        }

        /// <summary>
        /// In parts of New York City, some people feel REALLY STRONGLY about
        /// the hyphen in their house number. The numbering system makes sense,
        /// but the USPS address database doesn't support hyphens in the number field.
        /// To the USPS, the hyphen does not exist, but the DMM specifically does say
        /// that "if present, the hyphen should not be removed."
        /// </summary>
        [Fact]
        public void CanParseQueensStyleAddress()
        {
            Approvals.Verify(Parser.ParseAddress("123-465 34th St New York NY 12345"));
        }

        [Fact]
        public void CanParseTypicalAddressWithoutPunctuationAfterStreetLine()
        {
            Approvals.Verify(Parser.ParseAddress("1005 N Gravenstein Highway Sebastopol, CA 95472"));
        }

        [Fact]
        public void CanParseTypicalAddressWithPunctuation()
        {
            Approvals.Verify(Parser.ParseAddress("1005 N Gravenstein Highway, Sebastopol, CA 95472"));
        }

        [Fact]
        public void MatchSecondaryUnitAsWord()
        {
            var streets = new[] { "9999 RIVERSIDE DR WEST #8B", "9999 RIVERSIDE DR WEST SIDE LEFT", };
            var results = streets.Select(s => new { Text = s, Match = Regex.Match(s, AddressParser.Default.RangelessSecondaryUnitPattern, AddressParser.Default.MatchOptions) });
            Approvals.VerifyAll(results, "Match");
        }

        [Fact]
        public void ParseExampleAddresses()
        {
            var text = File.ReadAllText(PathUtilities.GetAdjacentFile("samples.txt"));
            var examples = text.Split(new[] {"\r\n", "\n"}, StringSplitOptions.RemoveEmptyEntries)
                .Where(s => !s.StartsWith("#", StringComparison.OrdinalIgnoreCase));
            bool[] normalize = { true, false };

            CombinationApprovals.VerifyAllCombinations(
                Parser.ParseAddress,
                x => Environment.NewLine + ((AddressParseResult)x).WritePropertiesToString(),
                examples,
                normalize);
        }

        [Fact]
        public void ParseWithoutNormalization()
        {
            Approvals.Verify(Parser.ParseAddress("999 West 89th Street Apt A New York NY 10024", false));
        }

        [Fact]
        public void VerifyAllSecondaryUnitPattern()
        {
            Approvals.Verify(AddressParser.Default.AllSecondaryUnitPattern);
        }

        [Fact]
        public void VerifyAllUnits()
        {
            Approvals.VerifyAll(AddressParser.Default.AllUnits);
        }

        [Fact]
        public void VerifyCityAndStatePattern()
        {
            Approvals.Verify(AddressParser.Default.CityAndStatePattern);
        }

        [Fact]
        public void VerifyDirectionalNames()
        {
            Approvals.VerifyAll(AddressParser.Default.DirectionalNames);
        }

        [Fact]
        public void VerifyDirectionalPattern()
        {
            Approvals.Verify(AddressParser.Default.DirectionalPattern);
        }

        [Fact]
        public void VerifyMatchOptions()
        {
            Approvals.Verify(AddressParser.Default.MatchOptions);
        }

        [Fact]
        public void VerifyPlacePattern()
        {
            Approvals.Verify(AddressParser.Default.PlacePattern);
        }

        [Fact]
        public void VerifyPostalBoxPattern()
        {
            Approvals.Verify(AddressParser.Default.PostalBoxPattern);
        }

        [Fact]
        public void VerifyRangedSecondaryUnitPattern()
        {
            Approvals.Verify(AddressParser.Default.RangedSecondaryUnitPattern);
        }

        [Fact]
        public void VerifyRangedUnits()
        {
            Approvals.VerifyAll(AddressParser.Default.RangedUnits);
        }

        [Fact]
        public void VerifyRangelessSecondaryUnitPattern()
        {
            Approvals.Verify(AddressParser.Default.RangelessSecondaryUnitPattern);
        }

        [Fact]
        public void VerifyRangelessUnits()
        {
            Approvals.VerifyAll(AddressParser.Default.RangelessUnits);
        }

        [Fact]
        public void VerifyStatePattern()
        {
            Approvals.Verify(AddressParser.Default.StatePattern);
        }

        [Fact]
        public void VerifyStates()
        {
            Approvals.VerifyAll(AddressParser.Default.StatesAndProvinces);
        }

        [Fact]
        public void VerifyStreetPattern()
        {
            Approvals.Verify(AddressParser.Default.StreetPattern);
        }

        [Fact]
        public void VerifySuffixes()
        {
            Approvals.VerifyAll(AddressParser.Default.StreetSuffixes);
        }

        [Fact]
        public void VerifySuffixPattern()
        {
            Approvals.Verify(AddressParser.Default.SuffixPattern);
        }

        [Fact]
        public void VerifyZipPattern()
        {
            Approvals.Verify(AddressParser.ZipPattern);
        }
    }
}