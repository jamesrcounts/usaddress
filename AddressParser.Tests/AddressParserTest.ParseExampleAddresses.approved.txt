[9999 ATLANTIC AVE BROOKLYN, N.Y. 99999,True] => 
AddressParseResult
{
	City: BROOKLYN
	Number: 9999
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: NY
	Street: ATLANTIC
	StreetLine: 9999 ATLANTIC AVE
	Suffix: AVE
	Zip: 99999
}

[9999 ATLANTIC AVE BROOKLYN, N.Y. 99999,False] => 
AddressParseResult
{
	City: BROOKLYN
	Number: 9999
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: N.Y.
	Street: ATLANTIC
	StreetLine: 9999 ATLANTIC AVE
	Suffix: AVE
	Zip: 99999
}

[P.O.Box 9999, Brea, CA 99999,True] => 
AddressParseResult
{
	City: BREA
	Number: 
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: CA
	Street: 
	StreetLine: POBOX 9999
	Suffix: 
	Zip: 99999
}

[P.O.Box 9999, Brea, CA 99999,False] => 
AddressParseResult
{
	City: Brea
	Number: 
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: CA
	Street: 
	StreetLine: P.O.Box 9999
	Suffix: 
	Zip: 99999
}

[3360 County Road F Tekamah NE 68061,True] => 
AddressParseResult
{
	City: TEKAMAH
	Number: 3360
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: NE
	Street: COUNTY ROAD F
	StreetLine: 3360 COUNTY ROAD F
	Suffix: 
	Zip: 68061
}

[3360 County Road F Tekamah NE 68061,False] => 
AddressParseResult
{
	City: Tekamah
	Number: 3360
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: NE
	Street: County Road F
	StreetLine: 3360 County Road F
	Suffix: 
	Zip: 68061
}

[623 NE 5th AVE Fort Lauderdale FL 33304,True] => 
AddressParseResult
{
	City: LAUDERDALE
	Number: 623
	Postdirectional: 
	Predirectional: NE
	SecondaryNumber: 
	SecondaryUnit: 
	State: FL
	Street: 5TH AVE
	StreetLine: 623 NE 5TH AVE FT
	Suffix: FT
	Zip: 33304
}

[623 NE 5th AVE Fort Lauderdale FL 33304,False] => 
AddressParseResult
{
	City: Lauderdale
	Number: 623
	Postdirectional: 
	Predirectional: NE
	SecondaryNumber: 
	SecondaryUnit: 
	State: FL
	Street: 5th AVE
	StreetLine: 623 NE 5th AVE Fort
	Suffix: Fort
	Zip: 33304
}

[Route 1  Gouldsboro ME 04607,True] => 
AddressParseResult
{
	City: 
	Number: 1
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: 
	Street: GOULDSBORO ME
	StreetLine: 1 GOULDSBORO ME
	Suffix: 
	Zip: 04607
}

[Route 1  Gouldsboro ME 04607,False] => 
AddressParseResult
{
	City: 
	Number: 1
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: 
	Street: Gouldsboro ME
	StreetLine: 1 Gouldsboro ME
	Suffix: 
	Zip: 04607
}

[3419 Avenue C Council Bluffs IA 51501,True] => 
AddressParseResult
{
	City: C COUNCIL BLUFFS
	Number: 3419
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: IA
	Street: AVENUE
	StreetLine: 3419 AVENUE
	Suffix: 
	Zip: 51501
}

[3419 Avenue C Council Bluffs IA 51501,False] => 
AddressParseResult
{
	City: C Council Bluffs
	Number: 3419
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: IA
	Street: Avenue
	StreetLine: 3419 Avenue
	Suffix: 
	Zip: 51501
}

[65 Ginger Woods  Valley NE 68064,True] => 
AddressParseResult
{
	City: 
	Number: 65
	Postdirectional: NE
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: 
	Street: GINGER WOODS
	StreetLine: 65 GINGER WOODS VLY NE
	Suffix: VLY
	Zip: 68064
}

[65 Ginger Woods  Valley NE 68064,False] => 
AddressParseResult
{
	City: 
	Number: 65
	Postdirectional: NE
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: 
	Street: Ginger Woods 
	StreetLine: 65 Ginger Woods Valley NE
	Suffix: Valley
	Zip: 68064
}

[Rte 175 Blue Hill Falls Rd  Blue Hill ME 04614,True] => 
AddressParseResult
{
	City: BLUE HILL
	Number: 175
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: ME
	Street: BLUE HILL FALLS
	StreetLine: 175 BLUE HILL FALLS RD
	Suffix: RD
	Zip: 04614
}

[Rte 175 Blue Hill Falls Rd  Blue Hill ME 04614,False] => 
AddressParseResult
{
	City: Blue Hill
	Number: 175
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: ME
	Street: Blue Hill Falls
	StreetLine: 175 Blue Hill Falls Rd
	Suffix: Rd
	Zip: 04614
}

[1302 LUCERNE AVENUE  Lake Worth FL 33460,True] => 
AddressParseResult
{
	City: WORTH
	Number: 1302
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: FL
	Street: LUCERNE AVENUE
	StreetLine: 1302 LUCERNE AVENUE LK
	Suffix: LK
	Zip: 33460
}

[1302 LUCERNE AVENUE  Lake Worth FL 33460,False] => 
AddressParseResult
{
	City: Worth
	Number: 1302
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: FL
	Street: LUCERNE AVENUE 
	StreetLine: 1302 LUCERNE AVENUE Lake
	Suffix: Lake
	Zip: 33460
}

[RR 1 Box 3145  Sedgwick ME 04676,True] => 
AddressParseResult
{
	City: SEDGWICK
	Number: 1
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: ME
	Street: BOX 3145
	StreetLine: 1 BOX 3145
	Suffix: 
	Zip: 04676
}

[RR 1 Box 3145  Sedgwick ME 04676,False] => 
AddressParseResult
{
	City: Sedgwick
	Number: 1
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: ME
	Street: Box 3145
	StreetLine: 1 Box 3145
	Suffix: 
	Zip: 04676
}

[3360 County Road F Tekamah NE 68061,True] => 
AddressParseResult
{
	City: TEKAMAH
	Number: 3360
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: NE
	Street: COUNTY ROAD F
	StreetLine: 3360 COUNTY ROAD F
	Suffix: 
	Zip: 68061
}

[3360 County Road F Tekamah NE 68061,False] => 
AddressParseResult
{
	City: Tekamah
	Number: 3360
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: NE
	Street: County Road F
	StreetLine: 3360 County Road F
	Suffix: 
	Zip: 68061
}

[623 NE 5th AVE Fort Lauderdale FL 33304,True] => 
AddressParseResult
{
	City: LAUDERDALE
	Number: 623
	Postdirectional: 
	Predirectional: NE
	SecondaryNumber: 
	SecondaryUnit: 
	State: FL
	Street: 5TH AVE
	StreetLine: 623 NE 5TH AVE FT
	Suffix: FT
	Zip: 33304
}

[623 NE 5th AVE Fort Lauderdale FL 33304,False] => 
AddressParseResult
{
	City: Lauderdale
	Number: 623
	Postdirectional: 
	Predirectional: NE
	SecondaryNumber: 
	SecondaryUnit: 
	State: FL
	Street: 5th AVE
	StreetLine: 623 NE 5th AVE Fort
	Suffix: Fort
	Zip: 33304
}

[Route 1  Gouldsboro ME 04607,True] => 
AddressParseResult
{
	City: 
	Number: 1
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: 
	Street: GOULDSBORO ME
	StreetLine: 1 GOULDSBORO ME
	Suffix: 
	Zip: 04607
}

[Route 1  Gouldsboro ME 04607,False] => 
AddressParseResult
{
	City: 
	Number: 1
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: 
	Street: Gouldsboro ME
	StreetLine: 1 Gouldsboro ME
	Suffix: 
	Zip: 04607
}

[3419 Avenue C Council Bluffs IA 51501,True] => 
AddressParseResult
{
	City: C COUNCIL BLUFFS
	Number: 3419
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: IA
	Street: AVENUE
	StreetLine: 3419 AVENUE
	Suffix: 
	Zip: 51501
}

[3419 Avenue C Council Bluffs IA 51501,False] => 
AddressParseResult
{
	City: C Council Bluffs
	Number: 3419
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: IA
	Street: Avenue
	StreetLine: 3419 Avenue
	Suffix: 
	Zip: 51501
}

[65 Ginger Woods  Valley NE 68064,True] => 
AddressParseResult
{
	City: 
	Number: 65
	Postdirectional: NE
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: 
	Street: GINGER WOODS
	StreetLine: 65 GINGER WOODS VLY NE
	Suffix: VLY
	Zip: 68064
}

[65 Ginger Woods  Valley NE 68064,False] => 
AddressParseResult
{
	City: 
	Number: 65
	Postdirectional: NE
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: 
	Street: Ginger Woods 
	StreetLine: 65 Ginger Woods Valley NE
	Suffix: Valley
	Zip: 68064
}

[1302 LUCERNE AVENUE  Lake Worth FL 33460,True] => 
AddressParseResult
{
	City: WORTH
	Number: 1302
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: FL
	Street: LUCERNE AVENUE
	StreetLine: 1302 LUCERNE AVENUE LK
	Suffix: LK
	Zip: 33460
}

[1302 LUCERNE AVENUE  Lake Worth FL 33460,False] => 
AddressParseResult
{
	City: Worth
	Number: 1302
	Postdirectional: 
	Predirectional: 
	SecondaryNumber: 
	SecondaryUnit: 
	State: FL
	Street: LUCERNE AVENUE 
	StreetLine: 1302 LUCERNE AVENUE Lake
	Suffix: Lake
	Zip: 33460
}

